File: ProgrammingAssignments/HLD/Java/TypeHead.java
class TypeHead {
    static class Trie {
        Trie[] children = new Trie[26];
        ArrayList<String> sugg = new ArrayList<String>();
    }
    private HashMap<String, Integer> searchTermFreq = new HashMap<String, Integer>();
    private Trie root = new Trie();

    public void incrementSearchTermFrequency(String search_term, int increment){
        int newFreq = searchTermFreq.getOrDefault(search_term, 0) + increment;
        searchTermFreq.put(search_term, newFreq);    
        int searchLength = search_term.length();
        Trie temp = root;
        for(int i=0;i<searchLength;++i) {
            int ch = search_term.charAt(i) - 'a';
            Trie child = temp.children[ch];
            if(child == null) {
                child = new Trie();
                child.sugg.add(search_term);
                temp.children[ch] = child;
            } else {
                boolean found = false;
                for(String str:child.sugg) {
                    if(str.equals(search_term)) {
                        found = true;
                        break;
                    }
                }
                if(found == false){
                    if(child.sugg.size() < 5) {
                        child.sugg.add(search_term);
                        sortSuggestions(child.sugg);
                   } 
                    else {
                        child.sugg.add(search_term);
                        sortSuggestions(child.sugg);
                        child.sugg.remove(5);
                    }
                }
            }
            temp = child;
        }
    }

    private void sortSuggestions(List<String> sugg) {
         sugg.sort( (s1, s2) -> {
                        //compare their freq
                        if(! searchTermFreq.get(s1).equals(searchTermFreq.get(s2))) {
                            return searchTermFreq.get(s2) - searchTermFreq.get(s1);
                        }
                        //if freq is same then lexicographical
                        return s2.compareTo(s1);
            });
    }

    public String[] findTopXSuggestion(String queryPrefix, int X) {
        Trie temp = root;
        for(int i=0;i<queryPrefix.length();++i) {
            int ch = queryPrefix.charAt(i) - 'a';
            Trie child = temp.children[ch];
            if(child == null) {
                temp = null;
                break;
            }
            temp = child;
        }

        String[] suggestions = new String[X];
        if(temp == null) {
            Arrays.fill(suggestions, "");
            return suggestions;
        } else {
            sortSuggestions(temp.sugg);
            int count = 0;
            for(int i=0;i< Math.min(X, temp.sugg.size()); ++i) {
                suggestions[i] = temp.sugg.get(i);
                ++count;
            }
            for(int i=count;i<X;++i) {
                suggestions[i] = "";
            }
            Arrays.sort(suggestions, (s1, s2) -> s1.compareTo(s2));
            return suggestions;
        }
    }
};

File: ProgrammingAssignments/HLD/TypeHead.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ProgrammingAssignments.HLD
{
    class TrieNode
    {
        public TrieNode[] Child = new TrieNode[26];
        public List<Tuple<int, string>> BestFive = new List<Tuple<int, string>>();
        public bool IsEnd;
    }

    class TypeHead
    {
        public TrieNode Root;
        private Dictionary<string, int> Freq = new Dictionary<string, int>();

        public TypeHead()
        {
            Root = GetNode();
            Freq.Clear();
        }

        private TrieNode GetNode()
        {
            TrieNode pNode = new TrieNode();
            pNode.IsEnd = false;
            for (int i = 0; i < 26; i++)
                pNode.Child[i] = null;
            return pNode;
        }
        public void IncrementSearchTermFrequency(string search_term, int increment)
        {
            TrieNode temp = Root;
            Freq[search_term] = Freq.GetValueOrDefault(search_term, 0) + increment;
            for (int i = 0; i < search_term.Length; i++)
            {
                int chr_val = search_term[i] - 'a';
                if (temp.Child[chr_val] == null)
                {
                    temp.Child[chr_val] = GetNode();
                }
                temp = temp.Child[chr_val];
                temp.BestFive = temp.BestFive.OrderBy(x => x.Item1).ToList();
                bool found = false;
                for (int j = 0; j < temp.BestFive.Count; j++)
                {
                    if (temp.BestFive[j].Item2 == search_term)
                    {
                        found = true;
                        temp.BestFive[j] = new Tuple<int, string>(Freq[search_term], search_term);
                    }
                }
                if (!found && (temp.BestFive.Count < 5 || temp.BestFive[0].Item1 <= Freq[search_term]))
                {
                    if (temp.BestFive.Count < 5)
                    {
                        temp.BestFive.Add(new Tuple<int, string>(Freq[search_term], search_term));
                    }
                    else if (Freq[search_term] > temp.BestFive[0].Item1 || string.Compare(search_term, temp.BestFive[0].Item2) > 0)
                    {
                        temp.BestFive[0] = new Tuple<int, string>(Freq[search_term], search_term);
                    }
                }
                temp.BestFive = temp.BestFive.OrderBy(x => x.Item1).ToList();
            }
            temp.IsEnd = true;
        }

        public List<string> FindTopXSuggestion(string queryPrefix, int X)
        {
            TrieNode temp = Root;
            List<string> ans = new List<string>();
            for (int i = 0; i < queryPrefix.Length; i++)
            {
                int chr_val = queryPrefix[i] - 'a';
                if (temp.Child[chr_val] == null)
                {
                    temp = temp.Child[chr_val];
                    break;
                }
                temp = temp.Child[chr_val];
            }
            if (temp != null)
            {
                List<Tuple<int, string>> v = temp.BestFive;
                v = v.OrderBy(x => x.Item1).ToList();
                for (int j = v.Count - 1; j >= 0 && ans.Count < X; j--)
                {
                    ans.Add(v[j].Item2);
                }
            }
            while (ans.Count < X)
            {
                ans.Add("");
            }
            ans.Sort();
            return ans;
        }
    }
}
